src/main/java/com/team766/controllers/CSMSMPID.java

```diff
@@ -0,0 +1,214 @@
+package com.team766.controllers;
+
+import com.team766.config.ConfigFileReader;
+import com.team766.hal.RobotProvider;
+import com.team766.library.SetValueProvider;
+import com.team766.library.SettableValueProvider;
+import com.team766.library.ValueProvider;
+import com.team766.logging.Category; //Todo: ?
+import com.team766.logging.Logger;
+import com.team766.logging.Severity; // Todo: ?
+
+
+import com.revrobotics.SparkMaxAbsoluteEncoder;
+import com.revrobotics.SparkMaxPIDController;
+import com.revrobotics.CANSparkMax.ControlType;
+import com.revrobotics.CANSparkMax.IdleMode;
+import com.revrobotics.SparkMaxAbsoluteEncoder.Type;
+import java.io.IOError;
+import javax.swing.text.DefaultStyledDocument.ElementSpec;
+import com.ctre.phoenix.motorcontrol.NeutralMode;
+import com.revrobotics.CANSparkMax;
+import com.team766.framework.Mechanism;
+import com.team766.hal.MotorController;
+import com.team766.library.RateLimiter;
+
+
+public class CSMSMPID{
+	// The attributes of the class include references to the motor controller, SparkMax controller, PID controller, and absolute encoder
+	private MotorController mc1;
+	private CANSparkMax csm1;
+	private SparkMaxPIDController pid1;
+	private SparkMaxAbsoluteEncoder abs1;
+	//PID Related Variables
+	private static double dz1 = 0; 
+	private static double maxpos1 = 0;
+	private static double minpos1 = 0;
+	private static double maxvel1 = 0;
+	private static double maxaccel1 = 0;
+	private static double maxspeed1 = 0;
+	private static double minspeed1 = 0;
+	private static double currentPos = 0;
+	private static double combo;
+
+	//antigrav variable
+	private static double antiGravK;
+	//enum for which state the PID is in
+	public enum PIDSTATE{
+		PID,
+		OFF,
+		ANTIGRAV
+	}
+	//the state of the PID2
+	private PIDSTATE theState = PIDSTATE.OFF;
+	
+	//constructor for the class with no absolute encoder
+	public CSMSMPID(String configName) throws Exception{
+			//loggerCategory = Category.MECHANISMS;
+
+			try{
+				mc1 = RobotProvider.instance.getMotor(configName);
+				csm1 = (CANSparkMax)mc1;
+				pid1 = csm1.getPIDController();
+			}catch (IllegalArgumentException ill){
+				throw new Exception("Error instantiating the PID controller: " + ill);
```

Comment:
is there value in having this be a checked exception (instead of a `RuntimeException`)? do you expect callers to catch this exception and try to recover from it?

Reply:
right. I honestly didn't know the difference. Maybe this is a good 5 in 10 topic?